src/seq_mv/csr_matop_device.c

```diff
@@ -1912,6 +1912,102 @@ hypre_CSRMatrixDiagScaleDevice( hypre_CSRMatrix *A,
    return hypre_error_flag;
 }
 
+/*--------------------------------------------------------------------------
+ * adj_functor (Used in hypre_CSRMatrixPermuteDevice)
+ *--------------------------------------------------------------------------*/
+
+struct adj_functor : public thrust::unary_function<HYPRE_Int, HYPRE_Int>
+{
+   HYPRE_Int *ia_;
+
+   adj_functor(int *ia)
+   {
+      ia_ = ia;
+   }
+
+   __host__ __device__ HYPRE_Int operator()(HYPRE_Int i) const
+   {
+      return ia_[i + 1] - ia_[i];
+   }
+};
+
+/*--------------------------------------------------------------------------
+ * bii_functor (Used in hypre_CSRMatrixPermuteDevice)
+ *--------------------------------------------------------------------------*/
+
+struct bii_functor
+{
+   HYPRE_Int *p_, *ia_, *ib_, *rb_;
+
+   bii_functor(HYPRE_Int *p, HYPRE_Int *ia, HYPRE_Int *ib, HYPRE_Int *rb)
+   {
+      p_ = p;
+      ia_ = ia;
+      ib_ = ib;
+      rb_ = rb;
+   }
+
+   __host__ __device__ void operator()(HYPRE_Int i)
+   {
+      const HYPRE_Int r = rb_[i];
+      rb_[i] = ia_[p_[r]] + i - ib_[r];
+   }
+};
+
+/*--------------------------------------------------------------------------
+ * hypre_CSRMatrixPermuteDevice
+ *
+ * See hypre_CSRMatrixPermute.
+ *--------------------------------------------------------------------------*/
+
+HYPRE_Int
+hypre_CSRMatrixPermuteDevice( hypre_CSRMatrix  *A,
+                              HYPRE_Int        *perm,
+                              HYPRE_Int        *rqperm,
+                              hypre_CSRMatrix  *B )
+{
+   /* Input matrix */
+   HYPRE_Int         num_rows     = hypre_CSRMatrixNumRows(A);
+   HYPRE_Int         num_nonzeros = hypre_CSRMatrixNumNonzeros(A);
+   HYPRE_Int        *A_i          = hypre_CSRMatrixI(A);
+   HYPRE_Int        *A_j          = hypre_CSRMatrixJ(A);
+   HYPRE_Complex    *A_a          = hypre_CSRMatrixData(A);
+   HYPRE_Int        *B_i          = hypre_CSRMatrixI(B);
+   HYPRE_Int        *B_j          = hypre_CSRMatrixJ(B);
+   HYPRE_Complex    *B_a          = hypre_CSRMatrixData(B);
+
+   HYPRE_Int        *B_ii;
+
+   /* Build B_i */
+   HYPRE_THRUST_CALL(gather,
+                     perm,
+                     perm + num_rows,
+                     thrust::make_transform_iterator(thrust::make_counting_iterator(0), adj_functor(A_i)),
```

Comment:
@victorapm Have you checked if this works for Sycl and HIP? I didn't. Thanks!

Reply:
It works for HIP and this implementation is not meant to work with SYCL as is. In fact, it's under an `ifdef HYPRE_USING_CUDA || HYPRE_USING_HIP` block. Do you want to address the SYCL implementation in this PR?

Reply:
No, we don't need to. I didn't realized it was not under HYPRE_USING_SYCL. Thanks!